In [3]:
## Ortam Hazirlanmasi
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import statsmodels.api as sm
from sklearn.model_selection import cross_val_score
## Veri Setinin Yuklenmesi
df = pd.read_csv('Advertising.csv')
df = df.iloc[:, 1:len(df)]
df.head()

TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [4]:
X = df.drop('Sales', axis=1)
y = df[['Sales']]

In [5]:
y.head()

Sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [6]:
X.head()

TV  Radio  Newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

STATSMODEL

In [7]:
statsmodel = sm.OLS(y, X)
model = statsmodel.fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  Sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              3566.
Date:                Mon, 28 Apr 2025   Prob (F-statistic):                   2.43e-171
Time:                        17:30:28   Log-Likelihood:                         -423.54
No. Observations:                 200   AIC:                                      853.1
Df Residuals:                     197   BIC:                                      863.0
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
Radio          0.2222      0.009     23.595      0.000       0.204       0.241
Newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### **Açıklamalar ve Değerlendirmeler**

- **R-Kare**: Modelin bağımlı değişken (Sales) üzerindeki varyansın %98.2'sini açıkladığını gösteriyor. Bu çok yüksek bir değer ve modelin veriye oldukça iyi uyum sağladığını gösteriyor.
  
- **F-İstatistiği**: 3566, bu modelin genel olarak anlamlı olduğunu ve bağımsız değişkenlerin (TV, Radio, Newspaper) satış üzerinde istatistiksel olarak önemli bir etkisi olduğunu doğruluyor. F-İstatistiği için p-değeri çok küçük (2.43e-171), bu da modelin güvenilirliğini pekiştiriyor.

- **Katsayılar**:
  - **TV**: 0.0538, TV reklamlarının her bir birim artışı için satışın ortalama olarak 0.0538 birim arttığını gösterir.
  - **Radio**: 0.2222, Radyo reklamlarının her bir birim artışı için satışın ortalama olarak 0.2222 birim arttığını gösterir. Bu katsayı TV'ye göre daha yüksek olduğu için, radyo reklamlarının satışlar üzerinde daha güçlü bir etkisi olduğu söylenebilir.
  - **Newspaper**: 0.0168, Gazete reklamlarının her bir birim artışı için satışın ortalama olarak 0.0168 birim arttığını gösterir. Gazete reklamlarının etkisi diğer iki değişkene göre çok daha düşük.

- **p-değerleri**: 
  - **TV** ve **Radio** değişkenleri için p-değerleri 0.000, bu da bu değişkenlerin modelde anlamlı olduğunu gösteriyor.
  - **Newspaper** için p-değeri 0.013, bu da gazete reklamlarının da önemli bir etkisi olduğunu, ancak diğer iki değişkene göre daha düşük bir güvenle ifade eder.

- **Durbin-Watson İstatistiği**: 2.038, modelin hata terimlerinin bağımsız olduğunu gösterir. Bu değer 2'ye yakınsa, hata terimlerinin bağımsız olduğunu kabul edebiliriz.

- **Omnibus Testi** ve **Jarque-Bera Testi**: Bu testler, modelin hata terimlerinin normal dağılıp dağılmadığını test eder. Her ikisi de sırasıyla 0.050 ve 0.0296'lık p-değerleri ile modelin hata terimlerinin normal dağıldığını gösteriyor.

- **Çarpıklık (Skew)**: -0.232, modelin hata terimlerinin biraz sol çarpık olduğunu gösteriyor, ancak bu değer modelin güvenilirliğini fazla etkilemez.

- **Çentiklik (Kurtosis)**: 3.794, hata terimlerinin normal dağılımın çentikliğinden çok uzak olmadığını gösterir.

Bu sonuçlarla, TV, radyo ve gazete reklamlarının satışlar üzerinde önemli etkileri olduğunu, ancak radyo reklamlarının daha güçlü bir etkisi olduğunu görebiliriz. Model oldukça yüksek bir R-kare değerine sahip ve iyi bir performans sergiliyor.

**SCIKIT LEARN**

In [8]:
sreg = LinearRegression()
model = sreg.fit(X, y)

In [20]:
print("b0 = ",model.intercept_), 
print("b1 = ",model.coef_[0][0]), 
print("b2 = ",model.coef_[0][1]), 
print("b3 = ",model.coef_[0][2]), 
print("R^2 = ",model.score(X,y))

b0 =  [2.93888937]
b1 =  0.04576464545539761
b2 =  0.18853001691820456
b3 =  -0.00103749304247632
R^2 =  0.8972106381789522


Model = 2.93888937 + 0.04576465 * TV + 0.18853002 * Radio - 0.00103749 * Newspaper

In [27]:
gercek_y = y.head(10)
tahmin_y = pd.DataFrame(model.predict(X)[0:10])
gercek_x = X.head(10)

In [30]:
# Gerçek y, tahmin edilen y ve gerçek x verilerini birleştirelim
tablo = pd.concat([gercek_x, gercek_y, tahmin_y], axis=1)

# Kolonlara anlamlı isimler verelim
tablo.columns = ['TV', 'Radio', 'Newspaper', 'Gercek_Satis', 'Tahmin_Satis']

# Sonuçları gösterelim
tablo

TV  Radio  Newspaper  Gercek_Satis  Tahmin_Satis
0  230.1   37.8       69.2          22.1     20.523974
1   44.5   39.3       45.1          10.4     12.337855
2   17.2   45.9       69.3           9.3     12.307671
3  151.5   41.3       58.5          18.5     17.597830
4  180.8   10.8       58.4          12.9     13.188672
5    8.7   48.9       75.0           7.2     12.478348
6   57.5   32.8       23.5          11.8     11.729760
7  120.2   19.6       11.6          13.2     12.122953
8    8.6    2.1        1.0           4.8      3.727341
9  199.8    2.6       21.2          10.6     12.550849

In [31]:
tablo['hata'] = tablo['Gercek_Satis'] - tablo['Tahmin_Satis']
tablo

TV  Radio  Newspaper  Gercek_Satis  Tahmin_Satis      hata
0  230.1   37.8       69.2          22.1     20.523974  1.576026
1   44.5   39.3       45.1          10.4     12.337855 -1.937855
2   17.2   45.9       69.3           9.3     12.307671 -3.007671
3  151.5   41.3       58.5          18.5     17.597830  0.902170
4  180.8   10.8       58.4          12.9     13.188672 -0.288672
5    8.7   48.9       75.0           7.2     12.478348 -5.278348
6   57.5   32.8       23.5          11.8     11.729760  0.070240
7  120.2   19.6       11.6          13.2     12.122953  1.077047
8    8.6    2.1        1.0           4.8      3.727341  1.072659
9  199.8    2.6       21.2          10.6     12.550849 -1.950849

In [33]:
MSE = mean_squared_error(y, model.predict(X))
RMSE = np.sqrt(MSE)
MAE = np.mean(np.abs(y - model.predict(X)))
print("MAE = ", MAE)
print("MSE = ", MSE)
print("RMSE = ", RMSE)

MAE =  1.2520112296870685
MSE =  2.784126314510936
RMSE =  1.6685701407225697


MODEL TUNING (MODEL DOGRULAMA)

In [68]:
## Verinin Egitim ve Test Setlerine Bolunmesi
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=60)
print("Egitim Seti Boyutu: ", X_train.shape[0])
print("Test Seti Boyutu: ", X_test.shape[0])
print("Egitim Seti Boyutu: ", y_train.shape[0])
print("Test Seti Boyutu: ", y_test.shape[0])

Egitim Seti Boyutu:  160
Test Seti Boyutu:  40
Egitim Seti Boyutu:  160
Test Seti Boyutu:  40


In [69]:
sreg = LinearRegression()
model = sreg.fit(X_train, y_train) # Egitim seti ile model olusturuldu
print("b0 = ",model.intercept_),
print("b1 = ",model.coef_[0][0]),
print("b2 = ",model.coef_[0][1]),
print("b3 = ",model.coef_[0][2]),
print("R^2 = ",model.score(X_train,y_train))

b0 =  [3.17606082]
b1 =  0.04481075470401089
b2 =  0.19240025103730687
b3 =  -0.0077500022968894815
R^2 =  0.8924386252509565


In [70]:
# Egitim Hatasini hesaplayalim
MSE_Train = np.sqrt(mean_squared_error(y_train, model.predict(X_train))) # Egitim Hatasini hesaplayalim
# Test Hatasini hesaplayalim
MSE_Test = np.sqrt(mean_squared_error(y_test, model.predict(X_test))) # Test Hatasini hesaplayalim

print("Egitim Hatası: ", MSE_Train)
print("Test Hatası: ", MSE_Test)

Egitim Hatası:  1.6884698315130209
Test Hatası:  1.6276538859244956


## **K-Katli Cross Validation (MANTIKLI OLAN, Yapmam Gereken)**

In [ ]:
cross_val_score(model, X_train, y_train, cv=10, scoring='neg_mean_squared_error') 

array([-1.06377064, -3.16977842, -3.99765198, -3.8148933 , -2.48057801,
       -1.80931618, -5.40424207, -1.38967391, -7.32266265, -1.2886141 ])

In [ ]:
cross_val_score(model, X_test, y_test, cv=10, scoring='neg_mean_squared_error') 

array([-0.99906182, -4.82478553, -2.24351612, -3.33448381, -5.19171441,
       -1.34235518, -1.09617624, -2.77859369, -1.03463212, -1.25002457])

In [79]:
# MSE Train
MSE_Train_All = np.mean(-cross_val_score(model, X_train, y_train, cv=10, scoring='neg_mean_squared_error')) 

In [80]:
# MSE Test
MSE_Test_All = np.mean(-cross_val_score(model, X_test, y_test, cv=10, scoring='neg_mean_squared_error')) 

In [82]:
# RMSE Train
RMSE_Train = np.sqrt(np.mean(-cross_val_score(model, X_train, y_train, cv=10, scoring='neg_mean_squared_error')))

In [83]:
# RMSE Test
RMSE_Test = np.sqrt(np.mean(-cross_val_score(model, X_test, y_test, cv=10, scoring='neg_mean_squared_error')))

In [84]:
# Genel Hata
RMSE_Genel = np.sqrt(np.mean(-cross_val_score(model, X, y, cv=10, scoring='neg_mean_squared_error')))

In [86]:
print("MSE Train: ", MSE_Train_All)
print("MSE Test: ", MSE_Test_All)
print("RMSE Train: ", RMSE_Train)
print("RMSE Test: ", RMSE_Test)
print("RMSE Genel: ", RMSE_Genel)
# Modelin Test Seti Üzerindeki Performansı

MSE Train:  3.174118126536387
MSE Test:  2.4095343493442014
RMSE Train:  1.7816054912736397
RMSE Test:  1.552267486403101
RMSE Genel:  1.749276312684338
